symcrypt/src/lib.rs

```diff
@@ -0,0 +1,145 @@
+use core::ffi::c_void;
+use std::mem;
+use std::ptr;
+use symcrypt_sys;
+
+pub const SHA256_RESULT_SIZE: usize = symcrypt_sys::SYMCRYPT_SHA256_RESULT_SIZE as usize;
+pub const SHA384_RESULT_SIZE: usize = symcrypt_sys::SYMCRYPT_SHA384_RESULT_SIZE as usize;
+
+pub struct SymCryptInit;
+impl SymCryptInit {
+    pub fn new() {
+        unsafe {
+            //symcrypt_sys::SymCryptInit();
+            // TODO: Find out why SymCryptInit() breaks on linux / windows (BREAKING)
+        }
+    }
+}
+
+pub struct Sha256State {
+    state: symcrypt_sys::_SYMCRYPT_SHA256_STATE,
+    is_dirty: bool,
+}
+
+impl Sha256State {
+    // Sha256State
+    pub fn new() -> Self {
+        let mut instance = Sha256State {
+            state: symcrypt_sys::_SYMCRYPT_SHA256_STATE::default(),
+            is_dirty: false,
+        };
+        unsafe {
+            symcrypt_sys::SymCryptSha256Init(&mut instance.state);
+        }
+        instance
+    }
+
+    pub fn append(&mut self, data: &[u8]) {
+        unsafe {
+            symcrypt_sys::SymCryptSha256Append(
+                &mut self.state,                    // pState
+                data.as_ptr(),                      // pbData
+                data.len() as symcrypt_sys::SIZE_T, // cbData
+            );
+        }
+        self.is_dirty = true;
+    }
+
+    pub fn result(&mut self, result: &mut [u8; SHA256_RESULT_SIZE]) {
+        unsafe {
+            symcrypt_sys::SymCryptSha256Result(&mut self.state, result.as_mut_ptr());
+        }
+        self.is_dirty = false;
+    }
+}
+
+impl Drop for Sha256State {
+    fn drop(&mut self) {
+        if self.is_dirty {
```

Comment:
We are talking about clearing 100+ memory bytes.

In about all cases, result would have been previously call where it will do the appropriate wipe already.

I thought you crypto guys cared about performance???

This is the low level rust code that will be called for every hash operation.

Reply:
Drop is only called when the object is deallocated (goes out of scope), right? If you're doing many hash operations, you can keep a single hash object alive and reuse it, and then you'd only incur one extra wipe. To me that feels worthwhile for the additional safety. (The code as-is appears correct to me; I'm more concerned about potential future changes that could forget to set this.) @samuel-lee-msft is more of an expert on perf, though, so I'd like to hear his input as well.

Reply:
I'm in two minds here - I agree with Mitch that the performant path is reusing a Sha256State; if you care about the performance of `drop` then you also care about the performance of `new`, and in `new` it's unavoidable that you're calling `symcrypt_sys::SYMCRYPT_SHA256_STATE::default()` (itself wiping state using Rust's write_bytes), and `SymCryptSha256Init` redundantly vs. reusing the state.

However, it does seem fairly simple to keep track of `is_dirty` correctly here, and it would remove a SymCryptWipe in most programs using Sha256State.

I would be minded to keep it simple until there is a clear need for optimization.

Reply:
Most applications won't be re-using this object after calling result(). However, they will be making multiple calls to append(). It they only need one append they would call the stateless sha256() function.

Note, I thought this was simple??  For sha256, 128 bytes to be wiped. For sha384, 224 bytes.

Reply:
OK we will remove this is_dirty optimization. Won't consider any other optimizations in this layer. If necessary can be added in the future without changing the rust application contract. Thanks.

Reply:
Sounds good, I will remove the is_dirty